revocation/crl/fetcher/fetcher.go

```diff
@@ -0,0 +1,162 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fetcher provides Fetcher interface and its implementation to fetch
+// CRL from the given URL
+package fetcher
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/revocation/crl/cache"
+)
+
+// MaxCRLSize is the maximum size of CRL in bytes
+//
+// CRL examples: https://chasersystems.com/blog/an-analysis-of-certificate-revocation-list-sizes/
+const MaxCRLSize = 32 * 1024 * 1024 // 32 MiB
+
+// Fetcher is an interface that specifies methods used for fetching CRL
+// from the given URL
+type Fetcher interface {
+	// Fetch retrieves the CRL from the given URL.
+	Fetch(ctx context.Context, crlURL string) (bundle *cache.Bundle, fromCache bool, err error)
+
+	// Download downloads the CRL from the given URL and saves it to the
+	// cache.
+	Download(ctx context.Context, crlURL string) (bundle *cache.Bundle, err error)
+}
```

Comment:
The caching mechanism should be transparent to the caller.
```suggestion
// Fetcher is an interface that specifies methods used for fetching CRL
// from the given URL
type Fetcher interface {
	// Fetch retrieves the CRL from the given URL.
	Fetch(ctx context.Context, uri string) (base, delta *x509.RevocationList, err error)
}
```

Reply:
Although delta CRL is not supported yet, we need to reserve that in the interface.

Reply:
Updated.